### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [zyearn/zaver#8]
Would you like to add more error handling for return values from functions like the following?
- [printf](http://pubs.opengroup.org/onlinepubs/9699919799/functions/printf.html "Print  formatted output") ⇒ [main](https://github.com/zyearn/zaver/blob/1a696ed276db5474b77d9a291c16e6704e8600df/src/zaver.c#L38)
- [pthread_cond_wait](http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_cond_wait.html "Wait on a condition") ⇒ [threadpool_worker](https://github.com/zyearn/zaver/blob/414abb7a656125b18a9f33a76725a8d02f2c9f94/src/threadpool.c#L182)

**Comment by zyearn** [comment]
@elfring it is a good advice to check pthread_cond_wait, I will fix this later. printf is not always be checked becauase even it fails, it will not affect the correctness of the program.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Unchecked return values") a bit more.
Would you like to detect every error situation as early as possible?

How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2FC.2B.2B "Tools for static code analysis") also for your software?

**Comment by zyearn** [comment]
Thks. I use splint to improve my code. If I apply all the advices output by splint, the code will be hard to read(a lot of return-value check...). So I made some trade-offs, some common usage adopt by most applications are remain the same in the code

**Comment by elfring** [comment]
>If I apply all the advices output by splint, the code will be hard to read(a lot of return-value check...).

* Would you like to fix the remaining open issues?
* How do you think about to integrate additional software design approaches?

**Comment by zyearn** [comment]
what do you mean by "integrate additional software design approaches"? can you give me an example?

**Comment by elfring** [comment]
Further approaches:
* Can C++ class libraries take care of the involved software design concerns?
* Are you interested to apply [aspect-oriented software development](http://aspectc.org/ "AspectC++")?
* How do you think about to encapsulate error detection and corresponding exception handling as [a reusable aspect](http://en.wikipedia.org/wiki/Aspect_%28computer_programming%29 "Combination of a pointcut with advice") in your software?

**Comment by zyearn** [comment]
I think it's better to be written in C, just like Nginx does. I will inspect the software design approaches in Nginx and implement it in Zaver in few codes.

**Comment by elfring** [comment]
Do you find information sources like the following useful?
* Article "[A Reusable Aspect for Memory Allocation Checking](https://web.archive.org/web/20090808093257/http://research.msrg.utoronto.ca/ACC/Tutorial#A_Reusable_Aspect_for_Memory_All "AspectC Tutorial")"
* Document "[Aspect-Oriented Software Development - An Introduction](https://web.archive.org/web/20130512025224/http://www.info.ucl.ac.be/~jbrichau/courses/introductionToAOSD.pdf "AOSD")" by Johan Brichau and Theo D'Hondt
* Paper "[On Typesafe Aspect Implementations in C++](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.125.2164 "AspectC++")" by Daniel Lohmann and Olaf Spinczyk

**Comment by zyearn** [comment]
Thanks for the information. I read the links, it is helpful.
I think AOP may increase the difficulty in understanding the system, especially for beginners, while zaver is primarily written for beginer developers to understand server programming. So maybe it will a future work

**Comment by elfring** [comment]
>I think AOP may increase the difficulty in understanding the system, …

There are some design trade-offs involved. Aspect-oriented software development can make some details easier while additional tools increase also specific risks.


>…, while zaver is primarily written for beginer developers to understand server programming.

How does such information fit to the short project description "another fast and efficient HTTP server"?

**Comment by zyearn** [comment]
> How does such information fit to the short project description "another fast and efficient HTTP server"?

There's no contradiction. It is another fast and efficient HTTP server, and at the meanwhile, I hope the code is compact and understandable enough even for beginners :)

**Comment by elfring** [comment]
Would you also like to show eventually how to avoid "beginner" mistakes around the implementation of efficient and robust servers?

**Comment by zyearn** [comment]
I have written an [artical](http://lifeofzjs.com/blog/2015/05/16/how-to-write-a-server/) about how to build a robust server from a toy server and how to handle the difficulty I met in coding, I am improving this artical with the development of the project. But it is written in Chinese, maybe latter I will translate it to English. For now I want to focus on adding some new features to the project.

**Comment by elfring** [comment]
You are also juggling with development priorities as usual. I am curious on how this software will evolve further.
